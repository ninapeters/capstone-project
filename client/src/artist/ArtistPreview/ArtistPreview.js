import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { useEffect } from 'react'
import AudioButton from '../../app/AudioButton'
import Button from '../../app/Button'
import { SaveIcon, SkipIcon } from '../../app/Icons/Icons'

ArtistPreview.propTypes = {
  artist: PropTypes.string.isRequired,
  songTitle: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  image: PropTypes.string,
  songUrl: PropTypes.string,
  updateTrack: PropTypes.func.isRequired,
  stopPlayingSong: PropTypes.func,
  togglePlayingSong: PropTypes.func,
  toggleCurrentSongId: PropTypes.func.isRequired,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
  saveSong: PropTypes.func.isRequired,
  savedSongs: PropTypes.array,
}

export default function ArtistPreview({
  artist,
  songTitle,
  id,
  image,
  songUrl,
  updateTrack,
  stopPlayingSong,
  togglePlayingSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
  saveSong,
  savedSongs,
}) {
  useEffect(() => {
    if (songUrl === null) {
      updateTrack()
    }
    toggleCurrentSongId(id)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [songUrl])

  if (songUrl === null) return <></>

  return (
    <>
      <Preview key={id}>
        <Cover src={image} alt="" />
        <Content>
          <ArtistName>{artist}</ArtistName>
          <Song>{songTitle}</Song>
        </Content>
        <ButtonWrapper>
          <Button isRedIcon id={id} onClick={skipSong} testId="skip">
            <SkipIcon />
          </Button>
          <AudioButton
            onClick={togglePlayingSong}
            isSongPlaying={isSongPlaying && currentSongId === id}
            currentSongId={currentSongId}
          />
          <Button
            id={id}
            onClick={() => saveSong(id)}
            disabled={savedSongs?.some((song) => song.id === id)}
            testId="save"
          >
            <SaveIcon />
          </Button>
        </ButtonWrapper>
      </Preview>
    </>
  )

  function skipSong() {
    stopPlayingSong()
    updateTrack()
  }
}

const Preview = styled.li`
  display: grid;
  flex: 1 0 100%;
  grid-template-columns: 100%;
  grid-template-rows: 54vh auto;
  height: 100vh;
  position: relative;
  row-gap: 46px;
  scroll-snap-align: start;
`
const Cover = styled.img`
  border-radius: 0 0 70px 70px;
  box-shadow: var(--shadow-dark);
  height: 54vh;
  object-fit: cover;
  width: 100%;
`
const Content = styled.section`
  padding: 0 22px;
`
const ArtistName = styled.span`
  color: var(--primary-regular);
  display: block;
  font-size: 1.2em;
  font-weight: 700;
  margin-bottom: 8px;
  text-transform: uppercase;
`
const Song = styled.span`
  color: var(--primary-regular);
  font-size: 1.6em;
  font-weight: 700;
  line-height: 1;
  text-transform: uppercase;
`
const ButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
  position: absolute;
  top: 36%;
  width: 100%;
`
