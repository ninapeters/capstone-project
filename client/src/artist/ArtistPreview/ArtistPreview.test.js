import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import ArtistPreview from './ArtistPreview'

describe('ArtistPreview', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        stopPlayingSong={() => {}}
        togglePlayingSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls updateTrack correctly', () => {
    const updateTrackMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={updateTrackMock}
        stopPlayingSong={() => {}}
        togglePlayingSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    const button = getByTestId('skip-button')
    user.click(button)
    expect(updateTrackMock).toHaveBeenCalled()
  })

  it('calls stopPlayingSong correctly', () => {
    const stopPlayingSongMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        stopPlayingSong={stopPlayingSongMock}
        togglePlayingSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    const button = getByTestId('skip-button')
    user.click(button)
    expect(stopPlayingSongMock).toHaveBeenCalled()
  })

  it('calls togglePlayingSong correctly', () => {
    const togglePlayingSongMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        stopPlayingSong={() => {}}
        togglePlayingSong={togglePlayingSongMock}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    const button = getByTestId('audio-button')
    user.click(button)
    expect(togglePlayingSongMock).toHaveBeenCalled()
  })

  it('calls toggleCurrentSongId with the correct id', () => {
    const toggleCurrentSongIdMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        stopPlayingSong={() => {}}
        togglePlayingSong={() => {}}
        toggleCurrentSongId={() => toggleCurrentSongIdMock('a')}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    const button = getByTestId('audio-button')
    user.click(button)
    expect(toggleCurrentSongIdMock).toHaveBeenCalledWith('a')
  })

  it('switches the appearance of play and pause button by changing the state isSongPlaying', () => {
    const { getByTitle, queryByTitle, rerender } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(getByTitle('play')).toBeInTheDocument()

    rerender(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={true}
        currentSongId="a"
        saveSong={() => {}}
        savedSongs={[]}
      />
    )
    expect(queryByTitle('play')).not.toBeInTheDocument()
    expect(getByTitle('pause')).toBeInTheDocument()
  })

  it('calls saveSong correctly', () => {
    const saveSongMock = jest.fn()
    const { getByTestId } = render(
      <ArtistPreview
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        image="https://testurl.com"
        songUrl="https://testurl.com"
        updateTrack={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
        saveSong={saveSongMock}
        savedSongs={[]}
      />
    )
    const button = getByTestId('save-button')
    user.click(button)
    expect(saveSongMock).toHaveBeenCalled()
  })
})
