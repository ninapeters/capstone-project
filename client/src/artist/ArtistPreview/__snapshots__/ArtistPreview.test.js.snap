// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistPreview renders correctly 1`] = `
.c7 {
  background: var(--white-transparent-min);
  border: none;
  border-radius: 50%;
  box-shadow: var(--shadow-light);
  fill: var(--primary-regular);
  height: 60px;
  padding: 16px;
  width: 60px;
}

.c7:focus {
  outline: 0;
}

.c6 {
  background: var(--white-transparent-min);
  border: none;
  border-radius: 50%;
  box-shadow: var(--shadow-light);
  fill: var(--cta-red);
  height: 60px;
  padding: 16px;
  width: 60px;
}

.c6:disabled {
  background: var(--spotify-green-transparent);
  fill: var(--white);
}

.c6:focus {
  outline: 0;
}

.c8 {
  background: var(--white-transparent-min);
  border: none;
  border-radius: 50%;
  box-shadow: var(--shadow-light);
  fill: var(--spotify-green);
  height: 60px;
  padding: 16px;
  width: 60px;
}

.c8:disabled {
  background: var(--spotify-green-transparent);
  fill: var(--white);
}

.c8:focus {
  outline: 0;
}

.c0 {
  display: grid;
  -webkit-flex: 1 0 100%;
  -ms-flex: 1 0 100%;
  flex: 1 0 100%;
  grid-template-columns: 100%;
  grid-template-rows: 54vh auto;
  height: 100vh;
  position: relative;
  row-gap: 46px;
  -webkit-scroll-snap-align: start;
  -moz-scroll-snap-align: start;
  -ms-scroll-snap-align: start;
  scroll-snap-align: start;
}

.c1 {
  border-radius: 0 0 70px 70px;
  box-shadow: var(--shadow-dark);
  height: 54vh;
  object-fit: cover;
  width: 100%;
}

.c2 {
  padding: 0 22px;
}

.c3 {
  color: var(--primary-regular);
  display: block;
  font-size: 1.2em;
  font-weight: 700;
  margin-bottom: 8px;
  text-transform: uppercase;
}

.c4 {
  color: var(--primary-regular);
  font-size: 1.6em;
  font-weight: 700;
  line-height: 1;
  text-transform: uppercase;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  position: absolute;
  top: 36%;
  width: 100%;
}

<li
  class="c0"
>
  <img
    alt=""
    class="c1"
    src="https://testurl.com"
  />
  <section
    class="c2"
  >
    <span
      class="c3"
    >
      Unprocessed
    </span>
    <span
      class="c4"
    >
      Real
    </span>
  </section>
  <div
    class="c5"
  >
    <button
      class="c6"
      data-testid="skip-button"
      style="transform: none;"
    >
      <svg>
        skip.svg
      </svg>
    </button>
    <button
      class="c7"
      data-testid="audio-button"
      style="transform: none;"
    >
      <svg
        title="play"
      >
        play.svg
      </svg>
    </button>
    <button
      class="c8"
      data-testid="save-button"
      style="transform: none;"
    >
      <svg>
        save.svg
      </svg>
    </button>
  </div>
</li>
`;
